daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/DFDLXFunctions.scala

```diff
@@ -29,6 +30,76 @@ import org.apache.daffodil.processors.unparsers.UState
 import org.apache.daffodil.processors.unparsers.UnparseError
 import org.apache.daffodil.util.Maybe.Nope
 import org.apache.daffodil.util.Maybe.One
+import passera.unsigned.{UInt, ULong}
+
+/**
+ * This is "logical" shift left, meaning it sign-extends a negative signed value.
+ * dfdl:leftShift for signed values does sign-extension, for logical (zero-filling) left shift you must use an unsigned type
+ *
+ * @param recipes
+ * @param argType
+ */
+case class DFDLXLeftShift(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes)  {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+
+    val shiftLong = arg2.getLong
+    val shift = shiftLong.toInt
+    val width = argType.asInstanceOf[PrimNumeric].width.get
+    Assert.invariant(shift >= 0)
+    if(shift>=width)
+      dstate.SDE("Left shift not supported for shift greater or equal to "+ width)
```

Comment:
Should we include the type and the passed in width in this error message? It should make it more obvious that different types have different width limitations. And if the shift width is calculated, it helps the user know what the value was. We should also use a format string. For example

```scala
dstate.SDE("Left shift not supported for shift greater or equal to %s for s%. Shift was %s", width, argType.globalQName, shift)
```